src/data_watcher.cpp

```diff
@@ -161,31 +166,133 @@ std::optional<std::vector<eventInfo>> DataWatcher::readEvents()
+        {
+            // The created dir is a child directory inside the configured data
+            // path add watch for the created child subdirectories.
+            fs::path subDirPath =
+                _watchDescriptors.at(std::get<wd>(receivedEventInfo)) /
+                std::get<basename>(receivedEventInfo);
+            createWatchers(subDirPath);
+            return true;
+        }
+        else if (_dataPathToWatch.string().starts_with(absCreatedPath.string()))
+        {
+            // Was monitoring exisitng parent path of the configured data path
+            // and a new file/directory got created inside it.
+
+            auto modifyWatchIfExpected = [this](const fs::path& entry) {
+                if (_dataPathToWatch.string().starts_with(entry.string()))
+                {
+                    // Created DIR is in the tree of the configured path.
+                    // Hence, Add watch for the created DIR and remove its
+                    // parent watch until the JSON configured DIR creates.
+                    if (_dataPathToWatch == entry)
+                    {
+                        // Add configured event masks if created DIR is the
+                        // configured path.
+                        addToWatchList(entry, _eventMasksToWatch);
+                    }
+                    else
+                    {
+                        addToWatchList(entry, _eventMasksIfNotExists);
+                    }
+
+                    if (auto parent =
+                            std::ranges::find_if(_watchDescriptors,
+                                                 [&entry](const auto& wd) {
+                        return (wd.second == entry.parent_path());
+                    });
+                        parent != _watchDescriptors.end())
+                    {
+                        removeWatch(parent->first);
+                    }
+                    return true;
+                }
+                else if (entry.string().starts_with(_dataPathToWatch.string()))
+                {
+                    // Created DIR is expected only, and is a child directory of
+                    // the configured path.
+                    // Hence add watch for created dirs and don't remove it's
+                    // parent as created DIRs are childs of the configured DIR.
+                    addToWatchList(entry, _eventMasksToWatch);
+                    return true;
+                }
+                return false;
+            };
+            auto monitoringPath =
+                _watchDescriptors.at(std::get<wd>(receivedEventInfo));
+            std::ranges::for_each(
+                fs::recursive_directory_iterator(monitoringPath),
+                modifyWatchIfExpected);
+            return true;
+        }
+    }
+    return false;
+}
+
+void DataWatcher::removeWatch(int wd)
+{
+    fs::path pathToRemove = _watchDescriptors.at(wd);
```

Comment:
I believe we should minimize repeated accesses to the container. Instead, we should retrieve the data once and reuse it where needed.

For example, in this case, we're repeatedly accessing the map solely for debug tracing.